### DISCUSSION TITLE
Using Service Role with Supabase in Next.js Backend - Which Approach is Valid?

### DISCUSSION THREAD

**Discussion by ganeshrvel** [supabase/supabase#30739] [Questions]
I'm new to Supabase.  
I'm working on a Next.js application with Supabase and want to handle everything through the Next.js backend using `service role`. I'm comparing two approaches and need help understanding which is valid/secure.

For both approaches, I start with:
```sql
revoke all on schema public from anon; 
revoke all on schema public from authenticated;
```

### Approach 1: Manual User ID Filtering
```typescript
import { createClient } from '@supabase/supabase-js';

const supabase = createClient(supabaseUrl, serviceRoleSecret, {
  auth: {
    persistSession: false,
    autoRefreshToken: false,
    detectSessionInUrl: false,
  },
});

// Manually filter by user_id in every query
const { data, error } = await supabase
  .from('accounts')
  .select('*')
  .eq('user_id', user.id); // Manually attach user_id
```

### Approach 2: Service Role with JWT and RLS
```typescript
// Get user session
import { createServerClient } from "@supabase/ssr";
import { cookies } from "next/headers";

export async function createClient() {
  const cookieStore = await cookies()
  
  return createServerClient(
    SUPABASE_URL,
    SUPABASE_ANON_KEY,
    {
      cookies: {
        getAll() {
          return cookieStore.getAll()
        },
        setAll(cookiesToSet) {
          try {
            cookiesToSet.forEach(({name, value, options}) =>
              cookieStore.set(name, value, options)
            )
          } catch {
          }
        },
      },
    }
  )
}

// Get session token
const { data: { session: sessionToken } } = await supabase.auth.getSession()

// Create service role client with session token
const supabase = createClient(supabaseUrl, serviceRoleSecret, {
  auth: {
    persistSession: false,
    autoRefreshToken: false,
    detectSessionInUrl: false,
  },
  global: {
    headers: {
      Authorization: `Bearer ${sessionToken}`
    }
  }
});

// In Supabase:
CREATE POLICY "service_role can access authenticated user data"
    ON public.your_table
    AS RESTRICTIVE
    FOR ALL
    TO service_role
    USING (auth.uid() = user_id);
```

### Questions:
1. Is Approach 2 even possible? Can you use `auth.uid() = user_id` with service role when sessionToken is passed?
   - I would appreciate if someone can point me to an article which talks about this setup.
2. Is there a better way to handle this? All examples I see use authenticated clients, but I want everything through the backend.
3. Is Approach 1 the only way here?

I'm trying to understand the proper way to handle this for a production application. Looking for insights on security implications and best practices.

Thanks in advance!

**Comment by GaryAustin1** [comment]
Approach 2 is not possible.  The authorization header is what decides if the role is service role.  You also cannot use SSR clients with service role as the cookie would replace the default role with a user session.

Approach 1 with a supabase-js is the main way.

You could potentially create a new Postgres role and mint your own JWT with the jwt secret on the fly with sub UUID and role as `server_role` and still use auth.uid() in RLS but with a `server_role` role that obeys RLS.  You would have to grant that role to authenticator and all your schemas, tables and functions.

**Reply by ganeshrvel** [reply]
Thanks for the clarification. I have a couple of follow-up questions:

When you mentioned: "`as server_role and still use auth.uid() in RLS but with a server_role role that obeys RLS`" 

1.  Could I create a JWT for this `server_role`, include my website user's ID in this JWT, and then have `auth.uid()` in Supabase resolve to the current user's ID (the user signed into my website)?

2. Could you point me to any resources or documentation about creating these custom JWTs on the fly, for implementing this approach?

@GaryAustin1

**Reply by GaryAustin1** [reply]
1) Yes, that is what I was suggesting.

You have to use a jwt library from your server side to mint the JWT.  I think you could probably copy the user jwt if you have one and change role claim to `server_role` and the rest would work with auth.uid() and auth.jwt().  They just would not have role 'authenticated'.  You could also just set up a jwt with role, sub and the few other claims needed.

https://supabase.com/docs/guides/auth/jwts#jwts-in-supabase describes the JWTs.

I don't know of a guide on minting the jwt with a library for Supabase, but if you search the web even with supabase in the search you would probably find examples.  I know users are minting JWTs for special server side cases.

I don't know how many do this.  I'm just throwing it out as a way as you asked for RLS.  Most just do filters with service_role.

**Reply by ganeshrvel** [reply]
This worked. Thanks!

**Reply by GaryAustin1** [reply]
Can you clarify if you went with minting the JWT or just service_role and filters.

**Reply by ganeshrvel** [reply]
I just used the service roles. Minted a jwt, sent it to supabase. It worked. The only concern is that I have to call the getUser() api on every server request to mint the jwt, which is affecting the performance.

**Reply by GaryAustin1** [reply]
Yes.   You would have to do that even if you went with service_role and did not mint and just used a filter also as you need to verify the jwt.
Technically you could verify the jwt though to your API with the same jwt library much faster than calling getUser().   The only thing you would not know is if the user is signed out or deleted after they got their jwt as the jwt sent to your API is good until it expires.

**Reply by ganeshrvel** [reply]
Thanks for the suggestion.

I think someone has made an attempt to solve this extra api call issue: https://github.com/butttons/supabase-auth-helpers

What do you think of it?

**Reply by GaryAustin1** [reply]
Sorry, I'm not going to search thru their repository to see what they do.  I assume as they have the jwt_secret they are going to verify the jwt is signed and not expired.  Which is what I suggested.  That is also done by Supabase here: https://supabase.com/docs/guides/database/postgres/custom-claims-and-role-based-access-control-rbac#accessing-custom-claims-in-your-application

**Comment by KKKircheff** [comment]
'use server';
import {createServerClient} from '@supabase/ssr';
import {cookies} from 'next/headers';


export async function createServiceClient() {
    const cookieStore = await cookies();
    return createServerClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_SUPABASE_SERVICE_KEY!,
        {
            cookies: {
                getAll() {
                    return [];
                },
            },
        }
    );
}



Here is the implementation how I use Supabase with service role that is simple and works perfect for me. Use only in case to edit tables that auth users don't have RLS permissions to. So tables with RLS set to access only with service_role

**Reply by GaryAustin1** [reply]
This is not correct.
You need to use supabase-js client and not SSR.  Also do not use cookies at all.  You don't want a user role to get into a service role initiated client.

Also the comment about set to access by service_role makes it sound like you need to set RLS to allow service_role.  You do not.  Service_role will never look at the RLS policies at all.  So  you can have RLS enabled an NO policy and service_role still gets thru.

**Reply by KKKircheff** [reply]
I use a service role to handle tables that are specific to users but restricted for authenticated and anon users in my backend. For example assigning credits or payment statuses to users or managing other admin information. For these operations, I create a service-role supabaseClient exclusively for the current admin task. Regular users interact with the app via the anon_key.

I will check supabase-js and I am open to solutions for instantiating a server-side client with the service role. By the way, I return an empty cookies array, so I don't think there will be an issue with cookies.

**Reply by GaryAustin1** [reply]
You just use supabase-js on the server (it is actually used by SSR).
If you set the cookie empty then you are "probably" OK, but the recommended way is just to use a plain supabase-js creatClient.
The issue comes in if the cookie has a user session, which the normal SSR process would have.

**Reply by KKKircheff** [reply]
Gary, thanks for suggestion. Just refactored the function to use supabase-js and it works fine on server side.

import {createClient} from '@supabase/supabase-js';

export function createServiceClient() {
    return createClient(
        process.env.NEXT_PUBLIC_SUPABASE_URL!,
        process.env.NEXT_SUPABASE_SERVICE_KEY!,
        {
            auth: {
                persistSession: false,
                autoRefreshToken: false,
                detectSessionInUrl: false,
            },
        }
    );
}
